compiler/codegen/library.cpp

```diff
@@ -243,37 +244,143 @@ static void setupPythonTypeMap() {
+
+    // Make import statement at top of .pyx file for chpl_library_init and
+    // chpl_library_finalize
+    fprintf(pyx.fptr, "from %s cimport chpl_library_init, ", pxdName);
+    fprintf(pyx.fptr, "chpl_library_finalize\n");
+
+    std::vector<FnSymbol*> moduleInits;
+    std::vector<FnSymbol*> exported;
+
+    fprintf(pyx.fptr, "from %s cimport ", pxdName);
+    bool first = true;
+    // Make import statement at top of .pyx file for exported functions
+    for_vector(FnSymbol, fn, functions) {
+      if (!isFunctionToSkip(fn)) {
+        if (fn->hasFlag(FLAG_EXPORT)) {
+          if (first) {
+            first = false;
+          } else {
+            fprintf(pyx.fptr, ", ");
+          }
+
+          // Module initialization functions get handled together, other
+          // exported functions will have their own definition.
+          if (fn->hasFlag(FLAG_MODULE_INIT)) {
+            // No need to rename the module init function, the user won't see it
+            fprintf(pyx.fptr, "%s", fn->cname);
+            moduleInits.push_back(fn);
+          } else {
+            // On import, rename the exported function so that we can use its
+            // original name in the Python module to avoid confusion.
+            fprintf(pyx.fptr, "%s as chpl_%s", fn->cname, fn->cname);
+            exported.push_back(fn);
+          }
         }
       }
+    }
+    fprintf(pyx.fptr, "\n\n");
 
-      pxdEnd();
-      gGenInfo->cfile = save_cfile;
+    // Necessary for using numpy types
+    fprintf(pyx.fptr, "import numpy\n");
+    fprintf(pyx.fptr, "cimport numpy\n\n");
+
+    makePYXSetupFunctions(moduleInits);
+
+    // Add Python wrapper for the exported functions, to translate the types
+    // appropriately
+    for_vector(FnSymbol, fn, exported) {
+      fn->codegenPython(false);
     }
-    // Don't "beautify", it will remove the tabs
-    closeLibraryHelperFile(&pxd, false);
+
+    gGenInfo->cfile = save_cfile;
   }
+  // Don't "beautify", it will remove the tabs
+  closeLibraryHelperFile(&pyx, false);
 }
 
+// Create definitions for chpl_setup and chpl_cleanup in the generated .pyx
+// file
+static void makePYXSetupFunctions(std::vector<FnSymbol*> moduleInits) {
+  GenInfo* info = gGenInfo;
+  FILE* outfile = info->cfile;
+
+  // Initialize the runtime.  chpl_setup should get called prior to using
+  // any of the exported functions
+  fprintf(outfile, "def chpl_setup():\n");
+  fprintf(outfile, "\tcdef char** args = ['%s']\n", libmodeHeadername);
+  fprintf(outfile, "\tchpl_library_init(1, args)\n");
+
+  // Initialize the included modules (continuation of chpl_setup definition)
+  for_vector(FnSymbol, fn, moduleInits) {
+    fprintf(outfile, "\t%s(1, 1)\n", fn->cname);
+  }
+  fprintf(outfile, "\n");
+
+  // Shut down the runtime and libraries.  chpl_cleanup should get called when
+  // the exported Chapel code is no longer needed
+  fprintf(outfile, "def chpl_cleanup():\n");
+  fprintf(outfile, "\tchpl_library_finalize()\n");
```

Comment:
Just curious, but it looks like chpl_library_init doesn't init modules but chpl_library_finalize does deinit modules. That seems odd. Am I misunderstanding?

Reply:
No, that's correct.  I was able to extend chpl_library_finalize to deinit modules because module initialization adds the deinit function to a list of functions to call, so the runtime only has to know that the list will be accurate instead of its contents.  We did discuss how to set up the library initialization functions to all get called by chpl_library_init back in May, but it required more work to implement and I haven't prioritized it yet.